doc/training_and_validation.md

```diff
@@ -0,0 +1,103 @@
+# Design: Training and Validation
+
+Validation is used to evaluate the trained model, like accuracy rating, overfitting, etc. In SQLFlow, automation is the core value. So SQLFlow proposed to do training and validation in one SQL. It requires splitting the user-specific data into two parts: training data and validation data correspond to two processes(train and validation) respectively. 
+
+Notice, we talk about the **train** process in this post.
+
+## Overall
+
+SQLFlow generates a temporary table following the user-specific dataset, trains and evaluates a model.
+
+<img src="/Users/weiguo/go/src/github.com/sql-machine-learning/sqlflow/doc/figures/training_and_validation.png" width="60%">
+
+## Generate a temporary dataset
+
+Splitting the training table into training data and validation data is the key point. We suppose SQLFlow are dealing with the following SQL to train an ML model:
+
+```sql
+​```
+SELECT col1, col2, col3
+FROM mytable
+TRAN ...
+​```
+```
+
+The data comes from the standard select part `SELECT col1, col2, col3 FROM mytable`, and let's say the query result looks like the following
+
+| col1   | col2   | col3   |
+| ------ | ------ | ------ |
+| <data> | <data> | <data> |
+| <data> | <data> | <data> |
+| <data> | <data> | <data> |
+| ...    |        |        |
+
+We want to split the result into 80% training data and 20% validation data.
+
+**We add a column sqlflow_random via RAND() and save the result to a temporary table.** Note the `RAND()` function returns a random number between 0 (inclusive) and 1. The result temporary table looks like the following.
+
+| temp_table |        |        |                |
+| ---------- | ------ | ------ | -------------- |
+| col1       | col2   | col3   | sqlflow_random |
+| <data>     | <data> | <data> | 0.3            |
+| <data>     | <data> | <data> | 0.9            |
+| <data>     | <data> | <data> | 0.5            |
+| ...        |        |        |                |
+
+We can generate the corresponding SQL using the following code template
```

Comment:
1. Maybe elaborate on how we generate the temporary table name. Something like `fmt.Sprintf("temp_table%d", rand.Int())`.
1. Which database does the temporary table belong to?

Reply:
1. I agree with generating an elaborate name for the temporary table. Because multi-users run SQLFlow with their own isolated dataset at the same time.

Reply:
2.1 For other databases, like MySQL. SQLFlow creates a temporary table with the attribute [TEMPORARY ](https://dev.mysql.com/doc/refman/8.0/en/create-temporary-table.html)
2.2 A `current project` must be specified by the user in Maxcompute.  SQLFlow saves the temporary table into the `current project`.

Let me add some details to this doc.